File: src/modules/mediation-manager.module.ts
import { MediationManagerModule } from '../utils/native';
import type {
  LastPageAdContent,
  MediationManagerEventListener,
  ShowAdCallbacks,
} from '../utils/types';
import { NativeEventEmitter } from 'react-native';
import { MediationManagerEvent } from '../utils/types';
import type { AdImpression } from '../utils/types';

type GenericEvent = {
  callbackId: string;
  data?: AdImpression | string;
};

export class MediationManager {
  private eventEmitter = new NativeEventEmitter(MediationManagerModule);
  private showAdListeners: Record<string, ShowAdCallbacks> = {};

  constructor() {
    this.eventEmitter.addListener('onShown', (e: GenericEvent) => {
      this.showAdListeners[e.callbackId]?.onShown?.(e.data as AdImpression);
    });

    this.eventEmitter.addListener('onShowFailed', (e: GenericEvent) => {
      this.showAdListeners[e.callbackId]?.onShowFailed?.(e.data as string);
      delete this.showAdListeners[e.callbackId];
    });

    this.eventEmitter.addListener('onClicked', (e: GenericEvent) => {
      this.showAdListeners[e.callbackId]?.onClicked?.();
    });

    this.eventEmitter.addListener('onComplete', (e: GenericEvent) => {
      this.showAdListeners[e.callbackId]?.onComplete?.();
      delete this.showAdListeners[e.callbackId];
    });

    this.eventEmitter.addListener('onClosed', (e: GenericEvent) => {
      this.showAdListeners[e.callbackId]?.onClosed?.();
      delete this.showAdListeners[e.callbackId];
    });
  }

  addListener = (
    event: MediationManagerEvent,
    listener: MediationManagerEventListener
  ) => {
    return this.eventEmitter.addListener(event, listener);
  };

  private createShowAdCallback = (props: ShowAdCallbacks) => {
    const id = Math.random().toString(36).slice(2, 7);
    this.showAdListeners[id] = props;
    return id;
  };

  setLastPageAdContent = async (params: LastPageAdContent) => {
    return MediationManagerModule.setLastPageAdContent(params);
  };

  loadInterstitial = async () => MediationManagerModule.loadInterstitial();

  isInterstitialReady = async (): Promise<boolean> =>
    MediationManagerModule.isInterstitialReady();

  showInterstitial = async (props: ShowAdCallbacks) => {
    return MediationManagerModule.showInterstitial(
      this.createShowAdCallback(props)
    );
  };

  loadRewardedAd = async () => MediationManagerModule.loadRewardedAd();

  isRewardedAdReady = async (): Promise<boolean> =>
    MediationManagerModule.isRewardedAdReady();

  showRewardedAd = async (props: ShowAdCallbacks) => {
    return MediationManagerModule.showRewardedAd(
      this.createShowAdCallback(props)
    );
  };

  enableAppReturnAds = async (props: ShowAdCallbacks) => {
    return MediationManagerModule.enableAppReturnAds(
      this.createShowAdCallback(props)
    );
  };

  disableAppReturnAds = async () =>
    MediationManagerModule.disableAppReturnAds();

  skipNextAppReturnAds = async () =>
    MediationManagerModule.skipNextAppReturnAds();

  loadAppOpenAd = async (isLandscape: boolean) =>
    MediationManagerModule.loadAppOpenAd(isLandscape);

  isAppOpenAdAvailable = async (): Promise<boolean> =>
    MediationManagerModule.isAppOpenAdAvailable();

  showAppOpenAd = async (props: ShowAdCallbacks) => {
    return MediationManagerModule.showAppOpenAd(
      this.createShowAdCallback(props)
    );
  };
}

File: ios/MediationManagerModule.m
//
//  CasModule.m
//  react-native-cas
//
//  Created by Lonely Astronaut on 20.09.23.
//

#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(MediationManagerModule, RCTEventEmitter)

RCT_EXTERN_METHOD(setLastPageAdContent:(NSDictionary*)content)

RCT_EXTERN_METHOD(loadInterstitial)
RCT_EXTERN_METHOD(isInterstitialReady:(RCTPromiseResolveBlock*) resolve
                  rejecter: (RCTPromiseRejectBlock*) reject)
RCT_EXTERN_METHOD(showInterstitial:(NSString*) callbackId)

RCT_EXTERN_METHOD(loadRewardedAd)
RCT_EXTERN_METHOD(isRewardedAdReady:(RCTPromiseResolveBlock*) resolve
                  rejecter: (RCTPromiseRejectBlock*) reject)
RCT_EXTERN_METHOD(showRewardedAd:(NSString*) callbackId)

RCT_EXTERN_METHOD(enableAppReturnAds:(NSString*) callbackId)
RCT_EXTERN_METHOD(disableAppReturnAds)
RCT_EXTERN_METHOD(skipNextAppReturnAds)

RCT_EXTERN_METHOD(loadAppOpenAd:(BOOL) isLandscape)
RCT_EXTERN_METHOD(isAppOpenAdAvailable:(RCTPromiseResolveBlock*) resolve
                  rejecter: (RCTPromiseRejectBlock*) reject)
RCT_EXTERN_METHOD(showAppOpenAd:(NSString*) callbackId)

@end

